File: auth-service/src/main/java/com/cicloparqueadero/auth/controller/AuthController.java
package com.cicloparqueadero.auth.controller;

import com.cicloparqueadero.auth.dto.*;
import com.cicloparqueadero.auth.security.AuthenticatedUser;
import com.cicloparqueadero.auth.service.AuthService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;

    @PostMapping("/register")
    public ResponseEntity<ApiResponse<AuthResponse>> register(@Valid @RequestBody RegisterRequest req) {
        AuthResponse data = authService.register(req);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok(data, "Usuario registrado exitosamente"));
    }

    @PostMapping("/login")
    public ResponseEntity<ApiResponse<AuthResponse>> login(@Valid @RequestBody LoginRequest req) {
        AuthResponse data = authService.login(req);
        return ResponseEntity.ok(ApiResponse.ok(data, "Inicio de sesión exitoso"));
    }

    @GetMapping("/profile")
    public ResponseEntity<ApiResponse<UserDTO>> profile(@AuthenticationPrincipal AuthenticatedUser principal) {
        UserDTO data = authService.getProfile(principal.getId());
        return ResponseEntity.ok(ApiResponse.ok(data));
    }

    @GetMapping("/users/count")
    public ResponseEntity<ApiResponse<Map<String, Long>>> count() {
        long total = authService.countUsers();
        return ResponseEntity.ok(ApiResponse.ok(Map.of("total", total)));
    }

    @GetMapping("/users")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ApiResponse<List<UserDTO>>> listUsers() {
        List<UserDTO> users = authService.getAllUsers();
        ApiResponse<List<UserDTO>> resp = ApiResponse.ok(users);
        resp.setTotal(users.size());
        return ResponseEntity.ok(resp);
    }

    @GetMapping("/validate")
    public ResponseEntity<ApiResponse<AuthenticatedUser>> validate(
            @AuthenticationPrincipal AuthenticatedUser principal) {
        return ResponseEntity.ok(ApiResponse.ok(principal));
    }
}

File: auth-service/src/controllers/auth.controller.js
const authService = require('../services/auth.service');

/**
 * Controlador de autenticación.
 * Solo coordina la entrada y salida HTTP.
 * La lógica vive en el servicio.
 */
class AuthController {
  async register(req, res, next) {
    try {
      const { nombre, correo, password, rol } = req.body;
      const result = await authService.register({ nombre, correo, password, rol });
      return res.status(201).json({
        success: true,
        message: 'Usuario registrado exitosamente',
        data: result,
      });
    } catch (error) {
      next(error);
    }
  }

  async login(req, res, next) {
    try {
      const { correo, password } = req.body;
      const result = await authService.login({ correo, password });
      return res.status(200).json({
        success: true,
        message: 'Inicio de sesión exitoso',
        data: result,
      });
    } catch (error) {
      next(error);
    }
  }

  async profile(req, res, next) {
    try {
      const user = await authService.getProfile(req.user.id);
      return res.status(200).json({
        success: true,
        data: user,
      });
    } catch (error) {
      next(error);
    }
  }

  async listUsers(req, res, next) {
    try {
      const users = await authService.getAllUsers();
      return res.status(200).json({
        success: true,
        data: users,
        total: users.length,
      });
    } catch (error) {
      next(error);
    }
  }

  async countUsers(req, res, next) {
    try {
      const total = await authService.countUsers();
      return res.status(200).json({
        success: true,
        data: { total },
      });
    } catch (error) {
      next(error);
    }
  }
}

module.exports = new AuthController();
